src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -1860,6 +2053,109 @@ open_common(int (*real_open)(const char *pathname, int oflags, ...), const char
+		if (rc != 0) {
+			DS_WARN(rc, "fetch_dfs_obj_handle() failed");
+			goto out_compatible;
+		}
+
+		/* Need to create a fake fd and associate with fd_kernel */
+		atomic_fetch_add_relaxed(&num_open, 1);
+		dfs_get_mode(dfs_obj, &mode_query);
+
+		/* regular file */
+		if (S_ISREG(mode_query)) {
+			rc = find_next_available_fd(NULL, &idx_fd);
+			if (rc)
+				goto out_compatible_release;
+			file_list[idx_fd]->dfs_mt    = dfs_mt;
+			file_list[idx_fd]->file      = dfs_obj;
+			file_list[idx_fd]->parent    = parent;
+			file_list[idx_fd]->st_ino    = FAKE_ST_INO(full_path);
+			file_list[idx_fd]->idx_mmap  = -1;
+			file_list[idx_fd]->open_flag = oflags;
+			file_list[idx_fd]->offset    = 0;
+			if (strncmp(full_path, "/", 2) == 0)
+				D_STRNDUP(file_list[idx_fd]->path, dfs_mt->fs_root, DFS_MAX_PATH);
+			else
+				D_ASPRINTF(file_list[idx_fd]->path, "%s%s", dfs_mt->fs_root, full_path);
+			if (file_list[idx_fd]->path == NULL) {
+				free_fd(idx_fd, false);
+				goto out_compatible;
+			}
+			strncpy(file_list[idx_fd]->item_name, item_name, DFS_MAX_NAME);
+			fd_fake = idx_fd + FD_FILE_BASE;
+		} else if (S_ISDIR(mode_query)) {
+			/* directory */
+			rc = find_next_available_dirfd(NULL, &idx_dirfd);
+			if (rc)
+				goto out_compatible_release;
+			dir_list[idx_dirfd]->dfs_mt   = dfs_mt;
+			dir_list[idx_dirfd]->fd       = idx_dirfd + FD_DIR_BASE;
+			dir_list[idx_dirfd]->offset   = 0;
+			dir_list[idx_dirfd]->dir      = dfs_obj;
+			dir_list[idx_dirfd]->num_ents = 0;
+			dir_list[idx_dirfd]->st_ino   = FAKE_ST_INO(full_path);
+			memset(&dir_list[idx_dirfd]->anchor, 0, sizeof(daos_anchor_t));
+			dir_list[idx_dirfd]->path = NULL;
+			dir_list[idx_dirfd]->ents = NULL;
+			if (strncmp(full_path, "/", 2) == 0)
+				full_path[0] = 0;
+
+			/* allocate memory for ents[]. */
+			D_ALLOC_ARRAY(dir_list[idx_dirfd]->ents, READ_DIR_BATCH_SIZE);
+			if (dir_list[idx_dirfd]->ents == NULL) {
+				free_dirfd(idx_dirfd);
+				goto out_compatible;
+			}
+			D_ASPRINTF(dir_list[idx_dirfd]->path, "%s%s", dfs_mt->fs_root, full_path);
+			if (dir_list[idx_dirfd]->path == NULL) {
+				free_dirfd(idx_dirfd);
+				goto out_compatible;
+			}
+			if (strnlen(dir_list[idx_dirfd]->path, DFS_MAX_PATH) >= DFS_MAX_PATH) {
+				DS_WARN(ENAMETOOLONG, "path is longer than DFS_MAX_PATH");
+				free_dirfd(idx_dirfd);
+				goto out_compatible;
+			}
+			fd_fake = idx_dirfd + FD_DIR_BASE;
+		} else {
+			/* not supposed to be here. If the object is neither a dir or a regular
+			 * file, fetch_dfs_obj_handle() should fail already. */
+			D_ASSERT(0);
+		}
+
+		/* add fd_kernel to hash table */
+		D_ALLOC_PTR(fd_ht_obj);
+		if (fd_ht_obj == NULL) {
+			if (fd_fake >= FD_DIR_BASE)
+				free_dirfd(idx_dirfd);
+			else
+				free_fd(idx_fd, false);
+			goto out_compatible;
+		}
+		fd_ht_obj->real_fd = fd_kernel;
+		fd_ht_obj->fake_fd = fd_fake;
+		rc = d_hash_rec_insert(fd_hash, &fd_ht_obj->real_fd, sizeof(int), &fd_ht_obj->entry,
+				       false);
+		D_ASSERT(rc == 0);
```

Comment:
assertion sounds concerning here.. why assert on error?

Reply:
d_hash_rec_insert always returns 0 if exclusive is false. Shall we just remove "D_ASSERT(rc == 0);"?

Reply:
Ashley previously suggested to add "D_ASSERT(rc == 0);". I think it is added to avoid coverity issues (not checking return value).

Reply:
I understand that if exclusive is false, the return value is always 0 currently. but if anyone changes it then this assertion is wrong. anyway, no need to change for now, but for coverity you can just add a D_ERROR or D_WARN messages on failure ..

Reply:
As the value of exclusive is false the return value can only be 0 so the use of D_ASSERT() here is correct.